File: Java/refactoring/src/test/java/za/co/entlect/refactoring1/service/CalculatorTest.java
package za.co.entlect.refactoring1.service;

import org.junit.Test;
import za.co.entelect.refactoring1.domain.BankAccount;
import za.co.entelect.refactoring1.domain.ChequeAccount;
import za.co.entelect.refactoring1.domain.MoneyMarketAccount;
import za.co.entelect.refactoring1.domain.SavingsAccount;
import za.co.entelect.refactoring1.exception.BankAccountException;
import za.co.entelect.refactoring1.service.Calculator;

import static junit.framework.Assert.assertEquals;

public class CalculatorTest {

    public static final long INITIAL_BALANCE = 1000L;

    private BankAccount savingsAccount;
    private BankAccount moneyMarketAccount;
    private BankAccount chequeAccount;

    private Calculator calculator = new Calculator();
    private BankAccount[] bankAccounts;

    @Test
    public void testAddBankAccount(){
        createTestAccounts(1000);
        assertEquals(0, calculator.countBanksAccounts());
        calculator.addBankAccount(chequeAccount);
        assertEquals(1, calculator.countBanksAccounts());
    }

    @Test
    public void testPositiveBalance() {
        createTestAccounts(INITIAL_BALANCE);

        calculator.calculate(savingsAccount);
        assertEquals(1050, savingsAccount.getBalanceInCents());

        calculator.calculate(chequeAccount);
        assertEquals(1040, chequeAccount.getBalanceInCents());

        calculator.calculate(moneyMarketAccount);
        assertEquals(1100, moneyMarketAccount.getBalanceInCents());
    }

    @Test
    public void testNegativeBalanceChequeAccount() {
        createTestAccounts(-500L);

        calculator.calculate(chequeAccount);
        assertEquals(-560, chequeAccount.getBalanceInCents());
    }

    @Test(expected = RuntimeException.class)
    public void testNegativeBalanceSavingsAccount() {
        createTestAccounts(-500L);
        calculator.calculate(savingsAccount);
    }

    @Test(expected = RuntimeException.class)
    public void testNegativeBalanceMoneyMarketAccount() {
        createTestAccounts(-500L);
        calculator.calculate(moneyMarketAccount);
    }

    @Test
    public void testUpdateBalance(){
        createTestAccounts(INITIAL_BALANCE);

        for (int i = 0; i < bankAccounts.length; i++) {
            calculator.updateBalance(bankAccounts[i], -500);
            assertEquals(500, bankAccounts[i].getBalanceInCents());

            calculator.updateBalance(bankAccounts[i], 1000);
            assertEquals(1500, bankAccounts[i].getBalanceInCents());
        }
    }

    @Test(expected = BankAccountException.class)
    public void testUpdateSavingsAccountBalanceFail(){
        createTestAccounts(INITIAL_BALANCE);
        calculator.updateBalance(savingsAccount, -INITIAL_BALANCE*2);
    }

    @Test(expected = BankAccountException.class)
    public void testUpdateMoneyMarketAccountBalanceFail(){
        createTestAccounts(INITIAL_BALANCE);
        calculator.updateBalance(moneyMarketAccount, -INITIAL_BALANCE*2);
    }

    private void createTestAccounts(long initialBalance) {
        savingsAccount = new SavingsAccount(initialBalance,
                BankAccount.SAVINGS_CREDIT_INTEREST_RATE, BankAccount.SAVINGS_DEBIT_INTEREST_RATE, BankAccount.SAVINGS_ACCOUNT_FEE);

        chequeAccount = new ChequeAccount(initialBalance,
                BankAccount.CHEQUE_CREDIT_INTEREST_RATE, BankAccount.CHEQUE_DEBIT_INTEREST_RATE, BankAccount.CHEQUE_ACCOUNT_FEE);

        moneyMarketAccount = new MoneyMarketAccount(initialBalance,
                BankAccount.MONEY_MARKET_CREDIT_INTEREST_RATE, BankAccount.MONEY_MARKET_DEBIT_INTEREST_RATE, BankAccount.MONEY_MARKET_ACCOUNT_FEE);

        bankAccounts = new BankAccount[]{savingsAccount, chequeAccount, moneyMarketAccount};
    }
}

File: C#/Refactoring/refactoring_exercise_1/test/za/co/entelect/refactoring1/service/CalculatorTest.cs

using NUnit.Framework;
using NUnit.Framework.Constraints;
using refactoring_exercise_1.za.co.entelect.refactoring1.domain;
using refactoring_exercise_1.za.co.entelect.refactoring1.exception;
using refactoring_exercise_1.za.co.entelect.refactoring1.service;

namespace refactoring_exercise_1.test.za.co.entelect.refactoring1.service
{
    [TestFixture]
    public class CalculatorTest {

        public const long InitialBalance = 1000L;

        private BankAccount _savingsAccount;
        private BankAccount _moneyMarketAccount;
        private BankAccount _chequeAccount;

        private readonly Calculator _calculator = new Calculator();
        private BankAccount[] _bankAccounts;

        [Test]
        public void TestAddBankAccount(){
            SavingsAccount savingsAccount = new SavingsAccount(1000L,
                BankAccount.SavingsCreditInterestRate, BankAccount.SavingsDebitInterestRate, BankAccount.SavingsAccountFee);
            Assert.That(_calculator.CountBanksAccounts(), new EqualConstraint(0));
            _calculator.AddBankAccount(savingsAccount);
            Assert.That(_calculator.CountBanksAccounts(), new EqualConstraint(1));
        }

        [Test]
        public void TestPositiveBalance() {
            CreateTestAccounts(InitialBalance);

            _calculator.Calculate(_savingsAccount);
            Assert.That(_savingsAccount.BalanceInCents, new EqualConstraint(1050));

            _calculator.Calculate(_chequeAccount);
            Assert.That(_chequeAccount.BalanceInCents, new EqualConstraint(1040));

            _calculator.Calculate(_moneyMarketAccount);
            Assert.That(_moneyMarketAccount.BalanceInCents, new EqualConstraint(1100));
        }

        [Test]
        public void TestNegativeBalanceChequeAccount() {
            CreateTestAccounts(-500L);

            _calculator.Calculate(_chequeAccount);
            Assert.That(_chequeAccount.BalanceInCents, new EqualConstraint(-560));
        }

        [Test]
        [ExpectedException( typeof (BankAccountException))]
        public void TestNegativeBalanceSavingsAccount() {
            CreateTestAccounts(-500L);
            _calculator.Calculate(_savingsAccount);
        }

        [Test]
        [ExpectedException( typeof (BankAccountException))]
        public void TestNegativeBalanceMoneyMarketAccount() {
            CreateTestAccounts(-500L);
            _calculator.Calculate(_moneyMarketAccount);
        }

        [Test]
        public void TestUpdateBalance()
        {
            CreateTestAccounts(InitialBalance);

            foreach (BankAccount account in _bankAccounts)
            {
                _calculator.CalculateBalance(account, -500);
                Assert.That(account.BalanceInCents, new EqualConstraint(500));

                _calculator.CalculateBalance(account, 1000);
                Assert.That(account.BalanceInCents, new EqualConstraint(1500));
            }
        }

        [Test]
        [ExpectedException( typeof (BankAccountException))]
        public void TestUpdateSavingsAccountBalanceFail(){
            CreateTestAccounts(InitialBalance);
            _calculator.CalculateBalance(_savingsAccount, -InitialBalance * 2);
        }

        [Test]
        [ExpectedException(typeof(BankAccountException))]
        public void TestUpdateMoneyMarketAccountBalanceFail(){
            CreateTestAccounts(InitialBalance);
            _calculator.CalculateBalance(_moneyMarketAccount, -InitialBalance * 2);
        }

        private void CreateTestAccounts(long initialBalance) {
            _savingsAccount = new SavingsAccount(initialBalance,
                BankAccount.SavingsCreditInterestRate, BankAccount.SavingsDebitInterestRate, BankAccount.SavingsAccountFee);

            _chequeAccount = new ChequeAccount(initialBalance,
                BankAccount.ChequeCreditInterestRate, BankAccount.ChequeDebitInterestRate, BankAccount.ChequeAccountFee);

            _moneyMarketAccount = new MoneyMarketAccount(initialBalance,
                BankAccount.MoneyMarketCreditInterestRate, BankAccount.MoneyMarketDebitInterestRate, BankAccount.MoneyMarketAccountFee);

            _bankAccounts = new[]{_savingsAccount, _chequeAccount, _moneyMarketAccount};
        }
    }
}
